From a216362a393976500b515efbe08e7d96c688600a Mon Sep 17 00:00:00 2001
From: Oliver Heger <oheger@apache.org>
Date: Tue, 9 Sep 2014 19:32:49 +0000
Subject: [PATCH] [CONFIGURATION-582] Fixed a StringIndexOutOfBoundsException.

Lines in properties files containing only whitespace are now handled
correctly when extracting comments.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/configuration/trunk@1623873 13f79535-47bb-0310-9956-ffa450edef68
---
 .../PropertiesConfigurationLayout.java        |  2 +-
 .../TestPropertiesConfigurationLayout.java    | 25 +++++++++++++++++++
 2 files changed, 26 insertions(+), 1 deletion(-)

diff --git a/src/main/java/org/apache/commons/configuration/PropertiesConfigurationLayout.java b/src/main/java/org/apache/commons/configuration/PropertiesConfigurationLayout.java
index ff54ba1dad..05e1af9e8f 100644
--- a/src/main/java/org/apache/commons/configuration/PropertiesConfigurationLayout.java
+++ b/src/main/java/org/apache/commons/configuration/PropertiesConfigurationLayout.java
@@ -724,7 +724,7 @@ static String trimComment(String s, boolean comment)
      */
     static String stripCommentChar(String s, boolean comment)
     {
-        if (s.length() < 1 || (isCommentLine(s) == comment))
+        if (StringUtils.isBlank(s) || (isCommentLine(s) == comment))
         {
             return s;
         }
diff --git a/src/test/java/org/apache/commons/configuration/TestPropertiesConfigurationLayout.java b/src/test/java/org/apache/commons/configuration/TestPropertiesConfigurationLayout.java
index 15a84c4b00..1ab3ee02da 100644
--- a/src/test/java/org/apache/commons/configuration/TestPropertiesConfigurationLayout.java
+++ b/src/test/java/org/apache/commons/configuration/TestPropertiesConfigurationLayout.java
@@ -660,6 +660,21 @@ public void testSetLineSeparatorInComments() throws ConfigurationException
                 + TEST_VALUE + lf);
     }
 
+    /**
+     * Tests whether a line with whitespace is handled correctly. This is
+     * related to CONFIGURATION-582.
+     */
+    @Test
+    public void testLineWithBlank() throws ConfigurationException
+    {
+        builder.addComment(TEST_COMMENT);
+        builder.addLine(" ");
+        builder.addProperty(TEST_KEY, TEST_VALUE);
+        layout.load(config, builder.getReader());
+        assertEquals("Wrong comment", TEST_COMMENT + CRNORM + " ",
+                layout.getCanonicalComment(TEST_KEY, false));
+    }
+
     /**
      * Helper method for filling the layout object with some properties.
      */
@@ -719,6 +734,16 @@ static class PropertiesBuilder
         /** A counter for varying the comment character. */
         private int commentCounter;
 
+        /**
+         * Adds a line verbatim to the simulated file.
+         *
+         * @param s the content of the line
+         */
+        public void addLine(String s)
+        {
+            buf.append(s).append(CR);
+        }
+
         /**
          * Adds a property to the simulated file.
          *
